docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,45 +2,268 @@
+Each row shows the imaginary queue and the process place in it. The proposer is the closest to the head of the queue, the rightmost validator. As priorities are updated, the validators move right in the queue. At the end of the round the proposer moves left as its priority is reduced after election. At the end of each round the sum of the priorities is 0.
+
+|Round Priority  | -2| -1| 0   |  1| 2   | 3 | 4 | 5 | Alg step
+|--------------- |---|---|---- |---|---- |---|---|---|--------
+|                |   |   |p1,p2|   |     |   |   |   |Initialized to 0
+|1               |   |   |     | p1|     | p2|   |   |A(i)+=VP(i)
+|                |   | p2|     | p1|     |   |   |   |A(p2)-= P
+|2               |   |   |     |   |p1,p2|   |   |   |A(i)+=VP(i)
+|                | p1|   |     |   |   p2|   |   |   |A(p1)-= P
+|3               |   | p1|     |   |     |   |   | p2|A(i)+=VP(i)
+|                |   | p1|     | p2|     |   |   |   |A(p2)-= P
+|4               |   |   |   p1|   |     |   | p2|   |A(i)+=VP(i)
+|                |   |   |p1,p2|   |     |   |   |   |A(p2)-= P
+
+The actual implementation does not create this queue, it only has to store the current accumulated proposer priority. 
+
+There are cases where the proposer selection is executed multiple times.
+(find a good example here)
+So we modify slightly the algorithm to handle this case. Note that times indicates the number of executions of the simpleProposerSelection() 
+
+    def singleProposerSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def ProposerSelection (vset, times):
+        for in in range(times):
+            singleProposerSelection(vset)
+
+### Validator Set Changes
+At each block height the validator set may change. Some of the changes have implications on the proposer selection.
+
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 4:
+
+
+Validator | p1| p2 
+----------|---| ---
+VP        | 4 | 3
+
+Let's also assume that at the last round R at height H before this change the proposer priorites were as shown in first row (H/R end). As it can be seen, the procedure could continue with round 1 at next height without changes as before. 
+
+|Round Priority| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| H/R end      |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| H+1/0        |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= P
+
+However, when a validator changes power from a high to a low value it may be stuck at the end of the queue for a long time if it happened to be there (e.g. it just proposed). This scenario is considered again after the New Validator section.
```

Comment:
Why is this important? He will eventually be selected in case period of stability is long enough?

Reply:
true but the problem may become severe, it is exactly issue described in https://github.com/tendermint/tendermint/issues/2960

Reply:
I will add a better example (the one from #2960)

Reply:
I am wondering if the idea of delaying new validator with setting priority to large number is indeed good idea. There are several weird cases this design opens. Maybe new validator can just be blocked of becoming proposer for some number of rounds/heights maybe proportional to its voting power. During this time his priority would stay to 0, so he will not mess up with the sum of priorities and also after it is effectively added he can join with priority 0.

Reply:
I was originally in favour of such an idea as well but we need to work out the details. 

What are the weird cases for the large negative initial value?

Reply:
I was thinking about the case where at the end of height H we have validator with significant voting power being removed from val set and new validator with much smaller voting power being added. Then new validator starts with negative voting power of old set, that is much smaller then the current total voting power. So he might stay at the end of the queue for significant amount of time.

Reply:
Changed the example that justifies the normalization